tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx

```diff
@@ -13,44 +13,96 @@
-   ULong64_t fEndEntry{};                      ///< The entry where the dataset processing should end (exclusive).
-   std::vector<std::string> fDefaultColumns{}; ///< A list of column names to process in the dataset.
+
+   struct REntryRange {
+      ULong64_t fStartEntry{0};
+      ULong64_t fEndEntry{std::numeric_limits<ULong64_t>::max()};
+      REntryRange() {}
+      REntryRange(ULong64_t endEntry) : fEndEntry(endEntry) {}
+      REntryRange(ULong64_t startEntry, ULong64_t endEntry)
+         : fStartEntry(startEntry),
+           fEndEntry(endEntry >= startEntry
+                        ? endEntry
+                        : throw std::logic_error("RDatasetSpec: fStartEntry cannot be larger than fEndEntry."))
+      {
+      }
+   };
+
    /**
     * A list of names of trees.
     * This list should go in lockstep with fFileNameGlobs, only in case this dataset is a TChain where each file
     * contains its own tree with a different name from the global name of the dataset.
     */
-   std::vector<std::string> fSubTreeNames{};
+   std::vector<std::string> fTreeNames{};
+
+   /**
+    * A list of file names.
+    * They can contain the globbing characters supported by TChain. See TChain::Add for more information.
+    */
+   std::vector<std::string> fFileNameGlobs{};
+
+   ULong64_t fStartEntry{}; ///< The entry where the dataset processing should start (inclusive).
+   ULong64_t fEndEntry{};   ///< The entry where the dataset processing should end (exclusive).
+
+   ROOT::Internal::TreeUtils::RFriendInfo fFriendInfo{}; ///< List of friends
+
+   RDatasetSpec(const std::string &treeName, const std::string &fileName, REntryRange entryRange = {})
+      : fTreeNames(std::vector<std::string>{treeName}), fFileNameGlobs(std::vector<std::string>{fileName}),
+        fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::string &treeName, const std::vector<std::string> &fileNames, REntryRange entryRange = {})
+      : fTreeNames(std::vector<std::string>{treeName}), fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry),
+        fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                REntryRange entryRange = {})
+      : fTreeNames(
+           fileNames.size() != treeNames.size() && treeNames.size() != 1
+              ? throw std::logic_error("RDatasetSpec exepcts either N trees and N files, or 1 tree and N files.")
+              : treeNames),
+        fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   void AddFriend(const std::string &treeName, const std::string &fileName)
+   {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
```

Comment:
I agree. This is better. I kept it currently differently, so that the current `AddFriend` overloads are "matching" the overloads of the spec, e.g. specifying  either 1. 1tree/1file; 2. 1tree/Nfiles; 3. Ntrees/Nfiles.

So I think we should settle down better overloads for `AddFriend` and I would adapt these vector emplacements accordingly.

I recall one of the overloads proposed was:
```cpp
void AddFriend(const std::string &treeName, const std::string &fileName, const std::string &treeAlias = "")
```
Does it makes sense to simply add `const std::string &treeAlias = ""` to each `AddFriend` overload?